### ISSUE TITLE
Simperium Not Saving 1-To-Many Relationship

### ISSUE THREAD

**Issue by lukegeiger** [Simperium/simperium-ios#430]
@jleandroperez 

I have a team object, that has many player objects... When I save the team none of the players get saved to simperium. They are indeed saved locally however...

There are no errors, in fact, simperium does not seem to even be trying to send up the Player object, when I turn on verbose logging enabled, All I see is 

Simperium marking object for sending more changes when ready (Player): 07383333b7714e058e62992059babf0c

**Comment by jleandroperez** [comment]
Hello there Luke,

Hope you're doing great!. Please, [set a breakpoint here](https://github.com/Simperium/simperium-ios/blob/develop/Simperium/SPWebSocketChannel.m#L191), and check the following flags:

```
p _indexing
p _authenticated
p (BOOL)processor.reachedMaxPendings
```

One of the above is causing the change to get enqueued, instead of processed. However, the change should get processed once the lib gets authenticated (OR) indexing finishes (OR) the pendings get cleared out.

Thanks!

**Comment by lukegeiger** [comment]
@jleandroperez Here are the results 

(lldb) p _indexing
(BOOL) $0 = NO
(lldb) p _authenticated
(BOOL) $1 = YES
(lldb) p processor.reachedMaxPendings
(BOOL) $2 = NO

**Comment by lukegeiger** [comment]
@jleandroperez Also, things to be working perfectly fine now, after I logged in on a new account... Is there anyway to precent this from happening in the future on my end?

**Comment by jleandroperez** [comment]
@lukegeiger i'm afraid there's nothing we can do, without more debug info. We'd need to reproduce it, and grab the status of those flags at the moment in which the relationship doesn't get sync'ed.

It could have been a networking glitch, heavy load (+200 pending changes), or maybe the lib was reindexing. But whatever it was, it should have recovered after a while (except if it was a networking issue!)

**Comment by lukegeiger** [comment]
@jleandroperez gotcha... So then the objects that weren't synced, will eventually get synced then? The only way a user would lose data is if they logged out when they had objects enqueued, correct?

**Comment by jleandroperez** [comment]
@lukegeiger right, signing out with pendings could cause that. 

That is, of course, as long as this wasn't caused by a bug. (Which without further debug info is quite impossible to determine). Nevertheless, the log message `Simperium marking object for sending more changes when ready` indicates that the entity is going to be effectively processed, when the conditions mentioned above change.

**Comment by lukegeiger** [comment]
@jleandroperez ok, thank  you for the quick responses. I'll make sure to keep an eye out for it and try to grab the logs.

**Comment by jleandroperez** [comment]
Awesome Luke, thank you!

**Comment by jleandroperez** [comment]
Let's close this one for now. Please, feel free to reopen if you're still experiencing this, thanks!

**Comment by lukegeiger** [comment]
@jleandroperez I am experiencing this issue again... What logs do you need me to send you?

**Comment by jleandroperez** [comment]
@lukegeiger verbose console logs please. Thanks!

**Comment by lukegeiger** [comment]
@jleandroperez ok this is all that gets printed out. Also, there is no bucket made on simperium.com either...

https://gist.github.com/lukegeiger/e9579d63d2711e686a52

**Comment by jleandroperez** [comment]
@lukegeiger please, set a breakpoint [here](https://github.com/Simperium/simperium-ios/blob/develop/Simperium/SPChangeProcessor.m#L647)  and:

- Check the contents of `object.ghost.memberData`
- Verify if the updated fields are effectively reflected in the `object` instance

Also, please, set a breakpoint [here](https://github.com/Simperium/simperium-ios/blob/develop/Simperium/SPMemberEntity.m#L104) and check what's the flow like, whenever a diff is calculated for one of the updated entities.

Thanks!

**Comment by lukegeiger** [comment]
@jleandroperez so the second breakpoint does not get hit, and the first breakpoint the object is my League object, when I think it should be the Conference object (Conference is the object that is not getting saved)

**Comment by lukegeiger** [comment]
@jleandroperez Also,

This is only occurring on my development Apps on Simperium. So for example, my live app on the app store is using version 0.7.5 and everything is working fine...

On my development app (same app just different Simperium ID and key) nothing works. I even jumped down to version 0.7.5 and it still didn't work.

So what is weird is that the same exact Simperium version is working live in the app store, but for whatever reason my development app is not.

**Comment by lukegeiger** [comment]
@jleandroperez after some more research, I am still just so confused on this whole thing... I can run the exact same code base and project pointing to the app store database and it works fine. But when I delete the app and reinstall pointing to the development database, absolutely nothing will save to Simperium without touching a single line of code. It does not even have to be a 1-To-Many, its ann entity.

**Comment by jleandroperez** [comment]
@lukegeiger just in case i'm not following... is this accurate? (Everything, with release 0.7.5):

- AppStore version works fine
- Develop build, mapped to Live Database works fine.
- Develop build, mapped to Development Database, breaks.

When your Develop build breaks, *nothing* syncs, not just 1 to many relationships.

Is that correct?

**Comment by lukegeiger** [comment]
@jleandroperez  Sorry for the confusion.
Everything is not working (no matter what build) for my development database.
Everything works fine on my live production database.

I even made a new account, and changed the API keys, and it still doesn't seem to want to work. This led me to believe maybe I changed something in my code, but whenever I change it back to the production database things work fine.

Just let me know what you need in terms of logs/debugging.

**Comment by jleandroperez** [comment]
@lukegeiger Would it be possible to run a quick test?.

Please, set your environment to build in *develop* mode, and map your AppStore database. Once ready, please, verify if sync'ing is working.

By *everything is not working* you mean 'entity addition' or 'entity update' right?. Would it be possible to get verbose logs for any of those OPs, with your develop db?.

Thank you!

**Comment by lukegeiger** [comment]
@jleandroperez sure, I am not sure what you mean by set the environment to develop mode, but map to app store. Can you clarify?

And yes I can not add any entities or update any entities. I will get the logs.

**Comment by jleandroperez** [comment]
@lukegeiger sure!. By develop mode i mean, pick the target you usually set to run debug builds (with extra symbols, and `DEBUG=1` flag).

The idea is: test exactly the same code that's breaking with your debug database, and map it to the live environment. If it just works, then we're dealing with an issue that doesn't live in the library itself.

Thanks!

**Comment by lukegeiger** [comment]
@jleandroperez Here are the logs.

Production
https://gist.github.com/lukegeiger/77f14444f34e552d6076

Development
https://gist.github.com/lukegeiger/54862b8dba9456c46e72

The only difference between these two builds are a matter of what database it points to. i.e setting up simperium in the app delegate.

**Comment by jleandroperez** [comment]
@lukegeiger is the Development change getting acknowledged?

```
2015-02-12 13:41:03.913 HM[75201:25721932] Simperium acknowledged change for ios-392b7cc4ef9043539b72f177dc4c51b2, cv=54dcf3bfa61f560f66563e56
```

How are you verifying this? secondary device mapped to develoment as well, right? (+ is anything showing up in those logs?).

**Comment by lukegeiger** [comment]
@jleandroperez no it is not getting acknowledged. It seems like almost overtime a different combination of logs gets printed out. for example now I am getting this 

    2015-02-12 14:08:44.996 HM[75628:25734575] Simperium marking object for sending more changes      when ready (League): f75c7a96f61e42f6b051980d5ec6273c
    2015-02-12 14:08:44.996 HM[75628:25734575] Simperium processing local object changes (User):   f6d8f017cd7145d7ad974b82c8d933a7
    2015-02-12 14:08:44.997 HM[75628:25734575] Simperium entity diff found 0 changed members
    2015-02-12 14:08:44.997 HM[75628:25734575] Simperium warning: no difference in call to sendChanges (User): f6d8f017cd7145d7ad974b82c8d933a7

**Comment by jleandroperez** [comment]
`Marking objects for sending more changes` message should only be displayed when:

- The library is indexing
- The websocket channels has not been authenticated
- There are over 200 changes that weren't acknowledged.

Please, [set a breakpoint here](https://github.com/Simperium/simperium-ios/blob/develop/Simperium/SPWebSocketChannel.m#L196) and check the status of the ivars.

Was the `User` entity effectively changed? (simperiumKey = `f6d8f017cd7145d7ad974b82c8d933a7`).

How are you testing this? two devices, and checking if the change gets through?

**Comment by lukegeiger** [comment]
@jleandroperez  here are the iVars

    (lldb) po _indexing
    YES
    (lldb) po _authenticated
    YES
    (lldb) po processor.reachedMaxPendings
    NO

I am testing this on one device at the moment. When I change from the Prod branch to the dev branch I delete the app and reinstall and do a fresh login.

The user was not changed.

**Comment by lukegeiger** [comment]
@jleandroperez Also, I should mention that one thing is actually saving... When the user signs up, I save a "Profile" object. And that gets synced with Simperium just fine, but everything after that does not work.

**Comment by jleandroperez** [comment]
@lukegeiger awesome, thanks for checking that!.

Apparently, (at least) one of your buckets is stuck indexing. This indexing process means that the client will retrieve a collection of 'last revisions per entity', and catch up with whatever is in the backend.

During the index process, sending new changes is forbidden (that's why the message says 'marking for later processing).

Problem is: the library should never get stuck indexing. *Sorry* about bothering you again with this, but would it be possible to capture *all* of the verbose logging output, starting from the moment in which you launch the app?.

Thanks again Luke!

**Comment by lukegeiger** [comment]
@jleandroperez of course, here are the logs from the moment I launch to after saving an object.
https://gist.github.com/lukegeiger/3087b1398990b21464f7

**Comment by lukegeiger** [comment]
@jleandroperez I just noticed this as well, when I switched devices, (testing on iPhone, now on the iPad) with the same account, Simperium seems to constantly print out the following logs.

https://gist.github.com/lukegeiger/959b302d6b57dc488392

**Comment by jleandroperez** [comment]
@lukegeiger *many* thanks for the logs!. I've isolated this flow (in the last gist):

```
2015-02-12 16:38:18.447 HM[78364:25804161] Simperium initializing websocket channel 6:{
    api = "1.1";
    "app_id" = "reservist-executions-1b5";
    clientid = "ios-0b0db6112e1447f9a6d92d398f3f4cd2";
    library = ios;
    name = LineupStats;
    token = adf80e0c36e047c1847df47268dbb59b;
    version = "0.7.5";
}
2015-02-12 16:38:18.564 HM[78364:25804161] Simperium () received "6:auth:admin@6thmanapps.com"
2015-02-12 16:38:18.565 HM[78364:25804161] Simperium client ios-0b0db6112e1447f9a6d92d398f3f4cd2 sending cv 6:cv:54ca848bd9b5bb0d6be4eaa4
2015-02-12 16:38:18.913 HM[78364:25804161] Simperium () received "6:cv:?"
2015-02-12 16:38:18.913 HM[78364:25804161] Simperium change version is out of date (LineupStats), re-indexing
2015-02-12 16:38:18.915 HM[78364:25804161] Simperium requesting index (LineupStats): 6:i::::500
2015-02-12 16:38:18.935 HM[78364:25804161] Simperium received index (LineupStats): {"index": []}
2015-02-12 16:38:18.935 HM[78364:25804161] Simperium processing 0 objects from index (LineupStats)
2015-02-12 16:38:18.936 HM[78364:25804161] Simperium finished processing all objects from index (LineupStats)
2015-02-12 16:38:18.937 HM[78364:25804161] Simperium client ios-0b0db6112e1447f9a6d92d398f3f4cd2 sending cv 6:cv:54ca848bd9b5bb0d6be4eaa4
```

That issue should be already fixed in #381. Mind updating to the latest stable release and running a quick test?.
